Achieve backwards compatibility between libcsp1 and libcsp2 ZMQ nodes

nml0 opened the pull request:
Related to https://github.com/libcsp/libcsp/issues/724

In libcsp-1 branch, ZMQ nodes don't send nor expect a CSP header with big endianness, they use host endianness instead. In libcsp2, ZMQ nodes do use big-endian (even when using `csp_conf.version = 1`), which creates incompatibility when mixing nodes.

Moreover, libcsp-1 branch sends an extra byte appended to the CSP header which contains the ZMQ destination address. This is not considered in libcsp2, where no ZMQ dest. address is sent (i.e., no rx filtering).

I think there are different ways to create this compatibility:

- we could just fix libcsp-1 to use big-endian / no ZMQ dest. address. However, I think changing libcsp-1 does not make sense now, as it's too old and many applications are already built against that version.
- [what I've tried to implement in this PR] modify libcsp2 such that we copy the mechanism in libcsp-1: we use ZMQ dst addr and host-endianness. This only applies to `csp_conf.version = 1`.

yashi commented on src/csp_id.c:
```diff
@@ -37,15 +37,30 @@
+					((uint32_t)(id1.sport) << CSP_ID1_SPORT_OFFSET) |
+					((uint32_t)(id1.flags) << CSP_ID1_FLAGS_OFFSET));
+	return id;
+}
+
+void csp_id1_deserialize(uint32_t id1, csp_packet_t * packet) {
+	/* Parse header */
+	packet->id.pri = (id1 >> CSP_ID1_PRIO_OFFSET) & CSP_ID1_PRIO_MASK;
+	packet->id.dst = (id1 >> CSP_ID1_DST_OFFSET) & CSP_ID1_DST_MASK;
+	packet->id.src = (id1 >> CSP_ID1_SRC_OFFSET) & CSP_ID1_SRC_MASK;
+	packet->id.dport = (id1 >> CSP_ID1_DPORT_OFFSET) & CSP_ID1_DPORT_MASK;
+	packet->id.sport = (id1 >> CSP_ID1_SPORT_OFFSET) & CSP_ID1_SPORT_MASK;
+	packet->id.flags = (id1 >> CSP_ID1_FLAGS_OFFSET) & CSP_ID1_FLAGS_MASK;
+}
+
 static void csp_id1_prepend(csp_packet_t * packet) {
 
-	/* Pack into 32-bit using host endian */
-	uint32_t id1 = (((uint32_t)(packet->id.pri) << CSP_ID1_PRIO_OFFSET) |
-					((uint32_t)(packet->id.dst) << CSP_ID1_DST_OFFSET) |
-					((uint32_t)(packet->id.src) << CSP_ID1_SRC_OFFSET) |
-					((uint32_t)(packet->id.dport) << CSP_ID1_DPORT_OFFSET) |
-					((uint32_t)(packet->id.sport) << CSP_ID1_SPORT_OFFSET) |
-					((uint32_t)(packet->id.flags) << CSP_ID1_FLAGS_OFFSET));
+	uint32_t id1 = csp_id1_serialize(packet->id);
```
Why did you need to move it to `csp_id1_serialize()`

nml0 replied:
Because with my current implementation, I need to access the serialized id from other files. However, given that:

> All of if (csp_conf.version == 1) { must be in csp_id1_prepend() or corresponding version 1 functions and not leaked out to examples or interface code.

(which seems perfectly fine to me, now that I think about it), this would not be needed. I will need to create new "CSP prepend / strip" functions for the ZMQ case though, because what I want to do in the end is just avoiding the big-endian/host-endian conversion. Do you agree with this?

yashi commented on src/interfaces/csp_if_zmqhub.c:
```diff
@@ -59,14 +73,14 @@ void * csp_zmqhub_task(void * param) {
 	zmq_driver_t * drv = param;
 	csp_packet_t * packet;
-	const uint32_t HEADER_SIZE = (csp_conf.version == 2) ? 6 : 4;
+	const uint32_t MSG_MIN_SIZE = (csp_conf.version == 2) ? 6 : 5; // CSP1 includes 1 extra byte for ZMQ dst addr
```
> CSP1 includes 1 extra byte for ZMQ dst addr

Wow. That was designed choise at that time? I wasn't aware of. that. (Again I don't use ZMQ)
I'd say the dst addr should be part of payload, not header.

nml0 replied:
I think so, let me know what you think. From` libcsp-1` branch:

```
/**
 * Interface transmit function
 * @param packet Packet to transmit
 * @return 1 if packet was successfully transmitted, 0 on error
 */
int csp_zmqhub_tx(const csp_route_t * route, csp_packet_t * packet) {

	zmq_driver_t * drv = route->iface->driver_data;

	const uint8_t dest = (route->via != CSP_NO_VIA_ADDRESS) ? route->via : packet->id.dst;

	uint16_t length = packet->length;
	uint8_t * destptr = ((uint8_t *) &packet->id) - sizeof(dest);
	memcpy(destptr, &dest, sizeof(dest));
	csp_bin_sem_wait(&drv->tx_wait, 1000); /* Using ZMQ in thread safe manner*/
	int result = zmq_send(drv->publisher, destptr, length + sizeof(packet->id) + sizeof(dest), 0);
	csp_bin_sem_post(&drv->tx_wait); /* Release tx semaphore */
	if (result < 0) {
		csp_log_error("ZMQ send error: %u %s\r\n", result, zmq_strerror(zmq_errno()));
	}

	csp_buffer_free(packet);

	return CSP_ERR_NONE;

}
```
`dstptr` is where the ZMQ destination address is saved. It's pointing to the byte previous to the `packet->id`.

yashi replied:
I just can't see why it's doing

```c
	uint8_t * destptr = ((uint8_t *) &packet->id) - sizeof(dest);
	memcpy(destptr, &dest, sizeof(dest));
```

It's taking address of `id`, and subtract `1`, size of `uint8_t`. Given that:

```c
typedef struct {
        uint8_t padding[CSP_PADDING_BYTES];
	uint16_t length;
	csp_id_t id;

	union {
		uint8_t data[0];
		uint16_t data16[0];
		uint32_t data32[0];
	};
} csp_packet_t;
```

it's overwriting a portion of `length` with `dest`? Am I reading it right?

nml0 replied:
That's right. It overwrites a preallocated byte with the ZMQ address (either the CSP DST address or via). Then, it's passed into ZMQ's backend by calling `zmq_send`, starting the Tx buffer that specific byte followed by the CSP packet.

Regarding the "why": I think this allows Rx filtering in the receiver side. In libcsp-1, in `csp_zmqhub_init_w_name_endpoints_rxfilter:` 
```
// subscribe to all 'rx_filters' -> subscribe to all packets, where the first byte (address/via) matches a rx_filter
for (unsigned int i = 0; i < rxfilter_count; ++i, ++rxfilter) {
	assert(zmq_setsockopt(drv->subscriber, ZMQ_SUBSCRIBE, rxfilter, 1) == 0);
}
```
which does not happen in libcsp2, where there is no (default) filtering.

yashi replied:
Ah, now I get it.

It just reusing packet buffer frrom `&packet->id - 1`. So that it can send 

```
 0 1234 567...
[d| id |......]
```
Where
- `d`: 1 byte destination address 
- `id`: 4 bytes ID (It's 8 bytes on 64 bits arch due to `unsigned int` in `csp_id_t`?!
- `....`: Payload data of `length` bytes

Thus, `length + sizeof(packet->id) + sizeof(dest)` in the third argument of `zmq_send()`.

yashi commented on src/interfaces/csp_if_zmqhub.c:
```diff
@@ -59,14 +73,14 @@ void * csp_zmqhub_task(void * param) {
 	zmq_driver_t * drv = param;
 	csp_packet_t * packet;
-	const uint32_t HEADER_SIZE = (csp_conf.version == 2) ? 6 : 4;
+	const uint32_t MSG_MIN_SIZE = (csp_conf.version == 2) ? 6 : 5; // CSP1 includes 1 extra byte for ZMQ dst addr
```
Would you please keep the name `HEADER_SIZE` to minimize the changes?

yashi commented on src/csp_id.c:
```diff
@@ -37,15 +37,30 @@
+	return id;
+}
+
+void csp_id1_deserialize(uint32_t id1, csp_packet_t * packet) {
```
Why do you need `packet`?

yashi commented on examples/zmqproxy.c:
```diff
@@ -59,14 +60,31 @@ static void * task_capture(void * ctx) {
 		csp_id_strip(packet);
 
+		if (csp_conf.version == 1)
+		{
```
We don't use hanging braces. Move it to the `if` line.

yashi commented on examples/zmqproxy.c:
```diff
@@ -59,14 +60,31 @@ static void * task_capture(void * ctx) {
+		if (csp_conf.version == 1)
+		{
+			// Revert the packet id endianness from host (used by ZMQ nodes in libcsp1) to big endian (libcsp2)
+			uint32_t id1 = csp_id1_serialize(packet->id);
+			id1 = htobe32(id1);
+			csp_id1_deserialize(id1, packet);
```
These must be handled by `csp_id_prepend()`

nml0 commented:
Building with systems `waf` and `meson` failed due to `unsigned int` casting - I fixed this.

yashi commented on src/interfaces/csp_if_zmqhub.c:
```diff
@@ -90,16 +100,23 @@ void * csp_zmqhub_task(void * param) {
+		const uint8_t* rx_data;
+		csp_zmq_get_msg_data(&msg, &rx_data, &datalen);
+
+		if (!rx_data) {
```
How can `rx_data` be `NULL`?

https://libzmq.readthedocs.io/en/latest/zmq_msg_data.html

yashi commented on src/csp_id.c:
```diff
@@ -56,6 +56,27 @@ static void csp_id1_prepend(csp_packet_t * packet) {
 	memcpy(packet->frame_begin, &id1, CSP_ID1_HEADER_SIZE);
 }
 
+static void csp_id1_prepend_host_endian(csp_packet_t * packet) {
```
This function is for "host" endian. but you have `csp_id1_strip_"big"_endian()`.

What is the endian of the ID for CSP1 ZMQ? It must be "little", no?

yashi commented:
> Building with systems `waf` and `meson` failed due to `unsigned int` casting - I fixed this.

I can't reproduce. How did you compile and what compiler and version are you using?
I can reproduce if I add `-Wconversion` to the buld. But then, many other files also produces warnings.

yashi commented on examples/zmqproxy.c:
```diff
@@ -15,6 +15,16 @@ const char * pub_str = "tcp://0.0.0.0:7000";
 char * logfile_name = NULL;
 FILE * logfile;
 
+static void csp_zmq_get_msg_data(zmq_msg_t* msg, const uint8_t** rx_data, unsigned int* datalen) {
```
Why do you have exact same function here and in `csp_if_zmqhub.c`?

yashi replied:
Made it to use the function in csp_if_zmqhub.c

nml0 commented:
> > Building with systems `waf` and `meson` failed due to `unsigned int` casting - I fixed this.
> 
> I can't reproduce. How did you compile and what compiler and version are you using? I can reproduce if I add `-Wconversion` to the buld. But then, many other files also produces warnings.

Sorry, I was not clear. I refer to a previous commit I did. I was building with cmake (worked fine) but the CI tests were failing for waf and meson. I now fixed that and squashed it into my commit.

yashi commented on src/csp_id.c:
```diff
@@ -56,6 +56,27 @@ static void csp_id1_prepend(csp_packet_t * packet) {
+
+	memcpy(packet->frame_begin, &id1, CSP_ID1_HEADER_SIZE);
+
+	// Add an extra byte before header: ZMQ dst = CSP dst
+	packet->frame_begin--;
+	*(packet->frame_begin) = (uint8_t)packet->id.dst;
+	packet->frame_length++;
```
Except the no endian conversions, these are extra step in addition to csp_id1_prepend().  Can you refactor?

yashi replied:
Created to function to add and remove the destination address.

yashi commented on src/interfaces/csp_if_zmqhub.c:
```diff
@@ -59,7 +69,7 @@ void * csp_zmqhub_task(void * param) {
 	zmq_driver_t * drv = param;
 	csp_packet_t * packet;
-	const uint32_t HEADER_SIZE = (csp_conf.version == 2) ? 6 : 4;
+	const uint32_t HEADER_SIZE = (csp_conf.version == 2) ? 6 : 5; // CSP1 includes 1 extra byte for ZMQ dst addr
```
I'd rather use another variable for the dst addr. The literal `6` and `4` should be replaced with predefined macros in another commit.

yashi replied:
Because we don't have any funciton to return header size, I haven't fixed this one yet. To do it, I need to first get packet, then call `csp_id_setup_rx()`, which make me reorder a few things and makes review harder. So, I'll do that later.

yashi reviewed (changes requested):
Thanks for the update. It's much better than the previous one. I have a few more request to fix before merging it.

yashi commented on src/csp_id.c:
```diff
@@ -37,7 +37,7 @@
 #define CSP_ID1_HEADER_SIZE 4
 
-static void csp_id1_prepend(csp_packet_t * packet) {
+static void csp_id1_prepend(csp_packet_t * packet, bool network_byte_order) {
```
My bad.

What I ment was to
- create a internal function to handle endianness of CSP ID, say, `__csp_id1_prepend()`
- call `__csp_id1_prepend()` from both `csp_id1_prepend()` and `csp_id1_zmq_prepend()`.

This way, we don't break the API.

yashi replied:
`csp_id1_prepend()` and `csp_id_strip()` are static functions. Thus, I left as is.  I've renamed `network_byte_order` to `csp1_fixup` to indicated that the parameter is only intended for ZQM interface quirk.

yashi commented on include/csp/interfaces/csp_if_zmqhub.h:
```diff
@@ -37,6 +37,21 @@ extern "C" {
+/**
+ * CSP1 header size.
+ */
+#define CSP1_HEADER_BASE_SIZE    4
```
We already have it as `CSP_ID1_HEADER_SIZE` and `CSP_ID2_HEADER_SIZE`

yashi replied:
will fix it later for simpler review.

yashi commented on include/csp/interfaces/csp_if_zmqhub.h:
```diff
@@ -37,6 +37,21 @@ extern "C" {
+/**
+ * ZMQ destination size (for libcsp1 backwards compatibility)
+ */
+#define ZMQ_DST_ADDR_SIZE    1
```
We don't use it outside of `src/interfaces/csp_if_zmqhub.c`. So we can move to `src/interfaces/csp_if_zmqhub.c`.

yashi replied:
done.

yashi reviewed (commented):
Thanks for the update. I wasn't clear enough with the previous review. Appologize.

yashi reviewed (commented):
Hi @nml0, Sorry for late response. I didn't re-run the workflow when you requested review.

Now I noticed that
1. some tests are failing, and
2. you haven't addressed what I pointed out in the last review (if I'm not mistaken).

If you are busy, I can take it from here, fix a few things and merge it. You are of cause welcome to fix the PR.
Please let me know what you think.

Thanks.

nml0 commented:
> If you are busy, I can take it from here, fix a few things and merge it. You are of cause welcome to fix the PR. Please let me know what you think.
> 
> Thanks.

Hello @yashi Sure, please go ahead with the changes. I'm afraid I didn't fully understand the request, as stated in the comments.

yashi commented:
I've updated your branch with my commits. I hope you don't mind that I added you as a sign-off, @nml0.

I still plan to:

- Add a cleanup commit in several places
- Add tests for v1 compatibility

It should still work the same way as what @nml0 prepared.

I mentioned this in the commit message, but I'm not entirely sure if it truly depends on the host byte order. All interfaces except ZMQ work without the fix. I might be misunderstanding something. Please let me know why endianness only affects the ZMQ interface.

If the little-endian header depends on the system's endianness, a big-endian system should theoretically work with stock libcsp v2. In that case, we'd need a compile-time condition to decide which endianness to use with the CSPv1 protocol on ZMQ.

Please review and test!

yashi commented:
@dimitrovand Do you have time to review and test?

dimitrovand commented:
I ran some basic tests locally on a little-endian machine, and it seems to work fine. Have you tested it on a big-endian machine or a mixed setup (one little-endian and one big-endian)?

yashi commented:
No I haven't. I'll do that in the following week

nml0 commented:
> If the little-endian header depends on the system's endianness, a big-endian system should theoretically work with stock libcsp v2. In that case, we'd need a compile-time condition to decide which endianness to use with the CSPv1 protocol on ZMQ.

Hi @yashi, I completely agree. I used mixed libcsp versions (libcsp-1 and libcsp-develop) in my tests on a little-endian machine. Therefore, we have a compatibility issue—the first one expects host (little) endian, and the second one expects big-endian. If we use a big-endian host, this issue wouldn't happen (I think, I haven't tested this) and we don't need to swap the endianness. I think we need to keep in mind the first byte, though: 

>  libcsp-1 branch sends an extra byte appended to the CSP header which contains the ZMQ destination address. This is not considered in libcsp2, where no ZMQ dest. address is sent (i.e., no rx filtering).

yashi commented:
OK, I've built and run libcsp-v1 on s390x, which is the only big-endian architecture currently available on Debian. I can confirm that, without byteswapping, it can communicate with the current version of libcsp via ZMQ. Additionally, the destination bytes must be preset.  
I can also confirm that the KISS interface works without any issues, either with or without the byteswapping ZMQ fix.

This means we need to specify the endianness used with the ZMQ interface, either at compile-time or run-time. Even though there may be only a small number of big-endian MCUs running libcsp-v1 (which is my wild guess), I believe we should default to big-endian. Users should be allowed to select the endianness for ZMQ only when CSPv1 is used. I suggest making this a compile-time flag, such as `CSP_V1_ZMQ_BIG_ENDIAN_FIXUP` or something similar. What do you think?

PS: Because `libzmq3-dev` is not yet ready for installation as multiarch, you need to build the s390x version of libcsp-v1 with `libzmq3-dev:s390x`, then replace it with `libzmq3-dev:amd64` to build the current version of libcsp. See [Debian bug #1092947](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1092947) for more details.

yashi commented:
preliminary change:

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 98f1a125..26ca06ac 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -39,6 +39,8 @@ option(CSP_BUFFER_ZERO_CLEAR "Zero out the packet buffer upon allocation" ON)
 option(CSP_ENABLE_PYTHON3_BINDINGS "Build Python3 binding" OFF)
 option(CSP_BUILD_SAMPLES "Build samples and examples by default" OFF)
 
+option(CSP_FIXUP_V1_ZMQ_LITTLE_ENDIAN "Use little-endian for ZMQ with CSPv1" ON)
+
 if(CMAKE_SYSTEM_NAME STREQUAL "Linux")
   set(CSP_POSIX 1)
 elseif(CMAKE_SYSTEM_NAME STREQUAL "Generic")
diff --git a/csp_autoconfig.h.in b/csp_autoconfig.h.in
index 1edb5aee..0c108c79 100644
--- a/csp_autoconfig.h.in
+++ b/csp_autoconfig.h.in
@@ -25,3 +25,5 @@
 
 #cmakedefine01 CSP_HAVE_LIBSOCKETCAN
 #cmakedefine01 CSP_HAVE_LIBZMQ
+
+#cmakedefine01 CSP_FIXUP_V1_ZMQ_LITTLE_ENDIAN
diff --git a/include/csp/csp_id.h b/include/csp/csp_id.h
index 41ec6687..fd6a60a7 100644
--- a/include/csp/csp_id.h
+++ b/include/csp/csp_id.h
@@ -15,8 +15,17 @@ unsigned int csp_id_get_max_port(void);
 
 int csp_id_is_broadcast(uint16_t addr, csp_iface_t * iface);
 
+#if (CSP_FIXUP_V1_ZMQ_LITTLE_ENDIAN)
 void csp_id_prepend_fixup_cspv1(csp_packet_t * packet);
 int csp_id_strip_fixup_cspv1(csp_packet_t * packet);
+#else
+static inline void csp_id_prepend_fixup_cspv1(csp_packet_t * packet) {
+	csp_id_prepend(packet);
+}
+static inline int csp_id_strip_fixup_cspv1(csp_packet_t * packet) {
+	return csp_id_strip(packet);
+}
+#endif
 
 #ifdef __cplusplus
 }
diff --git a/src/csp_id.c b/src/csp_id.c
index ae4c5fce..8ff0cdd9 100644
--- a/src/csp_id.c
+++ b/src/csp_id.c
@@ -202,6 +202,8 @@ int csp_id_strip(csp_packet_t * packet) {
 	}
 }
 
+#if (CSP_FIXUP_V1_ZMQ_LITTLE_ENDIAN)
+
 void csp_id_prepend_fixup_cspv1(csp_packet_t * packet) {
 	if (csp_conf.version == 2) {
 		csp_id2_prepend(packet);
@@ -219,6 +221,8 @@ int csp_id_strip_fixup_cspv1(csp_packet_t * packet) {
 	}
 }
 
+#endif
+
 int csp_id_setup_rx(csp_packet_t * packet) {
 	if (csp_conf.version == 2) {
 		csp_id2_setup_rx(packet);
```

yashi commented:
If someone want to corss-compile libcsp for s390x on Debian, here is how:

```
sudo dpkg --add-architecture s390x
sudo apt update
sudo apt install gcc-s390x-linux-gnu libzmq3-dev:s390x libsocketcan-dev:s390x
cat << EOF > s390x-linux-gnu.cmake
set(CMAKE_C_COMPILER s390x-linux-gnu-gcc)
set(CMAKE_C_COMPILER_ID GNU)
set(CMAKE_CXX_COMPILER s390x-linux-gnu-g++)
set(CMAKE_CXX_COMPILER_ID GNU)
set(ENV{PKG_CONFIG_LIBDIR} /usr/lib/s390x-linux-gnu/pkgconfig)
EOF
rm -rf builddir && cmake -B builddir -DCSP_BUILD_SAMPLES=1 -DCMAKE_TOOLCHAIN_FILE=s390x-linux-gnu.cmake && ninja -C builddir
```

nml0 commented:
> If someone wants to cross-compile libcsp for s390x on Debian, here’s how:
Do you use an emulator for executing this?

yashi commented:
> Do you use an emulator to execute this?

I just used QEMU.

```
qemu-s390x-static ./builddir/examples/csp_server -z localhost -a 1
```

yashi commented:
OK done adding compile time option and separated the destiation address reintroduction and byte-swapping. If this PR works for you guys, I'll merge it.  And after merging it, we can add some tests for compatibilities.

nml0 commented:
It works fine for the original setup that I used for finding this issue, which was:
* `libcsp-1`-based node (third party-app)
* `zmq_lib1_compatibility`-based ZMQ proxy
Hopefully, I'll have more time for testing during the weekend.

yashi commented:
OK. I'll wait and merge this next week then. Thanks.
